### ISSUE TITLE
Incorrect extracting of `{string}`

### ISSUE THREAD

**Issue by DDtKey** [cucumber-rs/cucumber-expressions#7]
In accordance with [cucumber-expressions](https://github.com/cucumber/cucumber-expressions#readme) - `{string}` should return only text between quotes. 

| Parameter Type  | Description |
| --------------- | ----------- |
| `{string}`      | Matches single-quoted or double-quoted strings, for example `"banana split"` or `'banana split'` (but not `banana split`). **Only the text between the quotes will be extracted. The quotes themselves are discarded.** Empty pairs of quotes are valid and will be matched and passed to step code as empty strings. |

But in current version there are quotes in String value

**Comment by ilslv** [comment]
@tyranron the trouble with this issue is that inside `cucumber` crate we map regex capturing groups one to one into callback parameters.

```rust
#[given(expr = "{param1} {param2} {param3}")]
fn callback(param1: i32, param2: i32, param3: i32) { ... }
```

So every parameter has to be represented with single capturing group and in case group didn't match we return empty string to avoid messing up the order.

We have to provide regex, that will match both `"text"` and `'text'` into single group containing `text` and I don't think this is possible with [`regex`][1] crate, as it would require remembering starting quote and matching until we find the same one.

Possible solutions to this issue
1. Return 2 groups with special names. Only 1 of them will match and handle another one inside `cucumber` crate based on a special name.
2. Require users to always escape `"` inside `'...'` and vice versa. 
3. Leave it as is. 

I like option 2 the most, because it simplifies the matching group, while not contradicting existing vague description of the language. But to be fair [actual implementation returns group containing 3 capturing groups: `("([^"\\]*(\\.[^"\\]*)*)"|'([^'\\]*(\\.[^'\\]*)*)')`][2] and `" ' "` is completely valid parameter. 

In case of Java, they are handling strings as [some kind of special way from what I can tell][3]. This is the only parameter consisting of 2 separate regex groups. 





[1]: https://docs.rs/regex/latest/regex/index.html
[2]: https://cucumber.github.io/cucumber-expressions/?advanced=1&expression=a%20boy%20named%20%7Bstring%7D&parameters=%5B%5D&step=a%20boy%20named%20%22Sue%22
[3]: https://github.com/cucumber/cucumber-expressions/blob/main/java/src/main/java/io/cucumber/cucumberexpressions/ParameterTypeRegistry.java#L39-L42

**Comment by tyranron** [comment]
@ilslv thank you for the clear explanation.

We cannot deviate much from what official implementation does, so the option 2 is not an option, unfortunately.

So, as the Java implementation does, we should handle `{string}` in a special way too on the `cucumber` side. Either with matching 2 capturing groups (as you've described), or by matching the whole, including quotes (as we do now), following by some post-procession (trimming quotes) in machinery before feeding it into the function.

**Comment by DDtKey** [comment]
If it's helpful, I'm currently using the following Wrapper-type for proper handling:

```rs
pub struct QuotedValue<T>(T);

impl<T: FromStr> FromStr for QuotedValue<T> {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(QuotedValue(
            T::from_str(&s[1..s.len() - 1]).map_err(|_| "Cannot extract quoted value")?,
        ))
    }
}

impl<T> std::ops::Deref for QuotedValue<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

```

But I guess this is not very useful for the library implementation 😅

**Comment by tyranron** [comment]
@DDtKey this is fixed now in 0.12.0 `cucumber` release.

Additionally, we have extended this case to custom parameters as well. So now, similar patterns where multiple capturing groups are required may be used for them too.

**Comment by DDtKey** [comment]
Thanks for the quick update! 💥
